grub-core/loader/multiboot_mbi2.c

```diff
@@ -295,29 +321,120 @@ grub_multiboot2_load (grub_file_t file, const char *filename)
 
       if (mld.relocatable)
 	{
-	  if (code_size > mld.max_addr || mld.min_addr > mld.max_addr - code_size)
+	  if (grub_slaunch_platform_type () == SLP_INTEL_TXT)
+	    {
+	      /*
+	       * We allocate the the binary together with page tables and MBI
+	       * to make one contiguous block for MLE. We have to align up to
+	       * PMR (2MB).
+	       */
+	      total_size = ALIGN_UP(code_size, MULTIBOOT_TAG_ALIGN); 
+	      total_size += grub_multiboot2_get_mbi_size();
+	      total_size = ALIGN_UP(total_size, GRUB_TXT_PMR_ALIGN);
+
+	      slparams->mle_size = total_size;
+
+	      slparams->mle_ptab_size = grub_txt_get_mle_ptab_size (total_size);
+	      slparams->mle_ptab_size = ALIGN_UP (slparams->mle_ptab_size, GRUB_TXT_PMR_ALIGN);
+
+	      total_size += slparams->mle_ptab_size;
+	      /* Do not go below GRUB_TXT_PMR_ALIGN. */
+	      if (mld.align > GRUB_TXT_PMR_ALIGN)
+		{
+		  mld.min_addr = (mld.min_addr > slparams->mle_ptab_size) ?
+				 (mld.min_addr - slparams->mle_ptab_size) : mld.align;
+		  mld.min_addr = ALIGN_UP (mld.min_addr, mld.align);
+		}
+	      else
+		{
+		  mld.min_addr = (mld.min_addr > slparams->mle_ptab_size) ?
+				 (mld.min_addr - slparams->mle_ptab_size) : GRUB_TXT_PMR_ALIGN;
+		  mld.min_addr = ALIGN_UP (mld.min_addr, GRUB_TXT_PMR_ALIGN);
+		  mld.align = GRUB_TXT_PMR_ALIGN;
+		}
+
+	    }
+	  else
+	   {
+	     total_size = code_size;
+	     slparams->mle_ptab_size = 0;
+	   }
+
+	  if (total_size > mld.max_addr || mld.min_addr > mld.max_addr - total_size)
 	    {
 	      grub_free (mld.buffer);
 	      return grub_error (GRUB_ERR_BAD_OS, "invalid min/max address and/or load size");
 	    }
 
 	  err = grub_relocator_alloc_chunk_align (grub_multiboot2_relocator, &ch,
-						  mld.min_addr, mld.max_addr - code_size,
-						  code_size, mld.align ? mld.align : 1,
+						  mld.min_addr, mld.max_addr - total_size,
+						  total_size, mld.align ? mld.align : 1,
 						  mld.preference, keep_bs);
 	}
       else
+	{
+	  if (grub_slaunch_platform_type () == SLP_INTEL_TXT)
+	    {
+	      if (!IS_ALIGNED(load_addr, GRUB_TXT_PMR_ALIGN)) 
```

Comment:
Just to confirm if I understand it correctly:
1. page tables are first in the allocated chunk, they **do not** have to be aligned to PMR
2. they are immediately followed (without any padding) by kernel that is aligned at least to PMR
3. kernel is followed by MBI, aligned to 8B or whatever the tags require
4. end of chunk is aligned up to PMR

Then MLE is set to everything between 2-4, including padding at the end, and that is measured to PCR17 by ACM, right?

Reply:
Ad .1. they have to (at least this is what Linux did).
Ad. 2. yes
Ad. 3. Yes
Ad. 4. Yes

> Then MLE is set to everything between 2-4, including padding at the end, and that is measured to PCR17 by ACM, right?

Yes